File: scripting/src/Api/Color.cs
using System;

namespace SEED;

/// <summary>
/// RGBA カラー（各成分 0.0〜1.0 の正規化 float）。スプライトの表示色や
/// カメラのクリアカラーなどに使う不変値型。
///
/// 定義済みカラー（White / Black / Red など）と、補間・乗算などの
/// よく使う操作を備える。不変（readonly）なので、変更したい場合は新しい値を代入する。
/// </summary>
public readonly struct Color : IEquatable<Color>
{
    /// <summary>赤成分（0.0〜1.0）。</summary>
    public readonly float r;
    /// <summary>緑成分（0.0〜1.0）。</summary>
    public readonly float g;
    /// <summary>青成分（0.0〜1.0）。</summary>
    public readonly float b;
    /// <summary>アルファ成分（0.0=透明〜1.0=不透明）。</summary>
    public readonly float a;

    /// <summary>RGBA を指定して生成する。アルファ省略時は不透明（1.0）。</summary>
    public Color(float r, float g, float b, float a = 1f)
    {
        this.r = r; this.g = g; this.b = b; this.a = a;
    }

    // ── 定義済みカラー ───────────────────────────────────────
    public static Color White       => new(1f, 1f, 1f, 1f);
    public static Color Black       => new(0f, 0f, 0f, 1f);
    public static Color Red         => new(1f, 0f, 0f, 1f);
    public static Color Green       => new(0f, 1f, 0f, 1f);
    public static Color Blue        => new(0f, 0f, 1f, 1f);
    public static Color Yellow      => new(1f, 1f, 0f, 1f);
    public static Color Cyan        => new(0f, 1f, 1f, 1f);
    public static Color Magenta     => new(1f, 0f, 1f, 1f);
    public static Color Gray        => new(0.5f, 0.5f, 0.5f, 1f);
    public static Color Transparent => new(0f, 0f, 0f, 0f);

    // ── 操作 ────────────────────────────────────────────────

    /// <summary>アルファだけを差し替えた新しいカラーを返す。</summary>
    public Color WithAlpha(float alpha) => new(r, g, b, alpha);

    /// <summary>a→b を t（0..1、クランプ）で線形補間する。</summary>
    public static Color Lerp(Color a, Color b, float t)
    {
        t = Mathf.Clamp01(t);
        return new Color(
            a.r + (b.r - a.r) * t,
            a.g + (b.g - a.g) * t,
            a.b + (b.b - a.b) * t,
            a.a + (b.a - a.a) * t);
    }

    // ── 演算子 ──────────────────────────────────────────────
    /// <summary>成分ごとの乗算（ティント合成）。</summary>
    public static Color operator *(Color x, Color y) => new(x.r * y.r, x.g * y.g, x.b * y.b, x.a * y.a);
    /// <summary>スカラー倍（アルファも含む）。</summary>
    public static Color operator *(Color c, float s) => new(c.r * s, c.g * s, c.b * s, c.a * s);
    public static bool operator ==(Color x, Color y) => x.Equals(y);
    public static bool operator !=(Color x, Color y) => !x.Equals(y);

    // ── 等価・表示 ───────────────────────────────────────────
    public bool Equals(Color other) => r == other.r && g == other.g && b == other.b && a == other.a;
    public override bool Equals(object? obj) => obj is Color c && Equals(c);
    public override int GetHashCode() => HashCode.Combine(r, g, b, a);
    public override string ToString() => $"RGBA({r:F3}, {g:F3}, {b:F3}, {a:F3})";
}

File: runtime/src/engine/structs/utils/color.rs
// ============================================================
//  color.rs — RGBA カラー型
// ============================================================

/// RGBA カラー（各チャンネル 0.0〜1.0）。
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Color {
    pub r: f32,
    pub g: f32,
    pub b: f32,
    pub a: f32,
}

impl Color {
    pub const fn new(r: f32, g: f32, b: f32, a: f32) -> Self {
        Self { r, g, b, a }
    }

    pub const fn rgb(r: f32, g: f32, b: f32) -> Self {
        Self { r, g, b, a: 1.0 }
    }

    pub fn with_alpha(self, a: f32) -> Self {
        Self { a, ..self }
    }

    /// `[r, g, b, a]` 配列に変換する（GPU バッファ・描画関数向け）。
    pub const fn to_array(self) -> [f32; 4] {
        [self.r, self.g, self.b, self.a]
    }

    // ── 基本色 ────────────────────────────────────────────────

    pub const BLACK: Self = Self::rgb(0.0, 0.0, 0.0);
    pub const WHITE: Self = Self::rgb(1.0, 1.0, 1.0);
    pub const RED: Self = Self::rgb(1.0, 0.0, 0.0);
    pub const GREEN: Self = Self::rgb(0.0, 1.0, 0.0);
    pub const BLUE: Self = Self::rgb(0.0, 0.0, 1.0);
    pub const YELLOW: Self = Self::rgb(1.0, 1.0, 0.0);
    pub const CYAN: Self = Self::rgb(0.0, 1.0, 1.0);
    pub const MAGENTA: Self = Self::rgb(1.0, 0.0, 1.0);
    pub const TRANSPARENT: Self = Self::new(0.0, 0.0, 0.0, 0.0);

    // ── ギズモ軸色（高彩度）──────────────────────────────────

    pub const GIZMO_X: Self = Self::rgb(1.0, 0.1, 0.1);
    pub const GIZMO_Y: Self = Self::rgb(0.1, 1.0, 0.1);
    pub const GIZMO_Z: Self = Self::rgb(0.1, 0.1, 1.0);
}

impl From<Color> for [f32; 4] {
    fn from(c: Color) -> Self {
        c.to_array()
    }
}

impl From<[f32; 4]> for Color {
    fn from(a: [f32; 4]) -> Self {
        Self::new(a[0], a[1], a[2], a[3])
    }
}
